### ISSUE TITLE
Request a feature to disable only specific ObservationHandlers

### ISSUE THREAD

**Issue by be-hase** [micrometer-metrics/micrometer#5467]
## Motivation

Micrometer Observation allows transparent measurement of both metrics and tracing. However, for legacy code that has already implemented metrics and tracing using older libraries such as Micrometer and Brave, transitioning entirely to Micrometer Observation can be challenging.

Take the example of [Lettuce](https://github.com/redis/lettuce): I have been using [MicrometerCommandLatencyRecorder](https://github.com/redis/lettuce/blob/main/src/main/java/io/lettuce/core/metrics/MicrometerCommandLatencyRecorder.java) for metrics and [BraveTracing](https://github.com/redis/lettuce/blob/main/src/main/java/io/lettuce/core/tracing/BraveTracing.java) for tracing. Now, I want to migrate to [io.lettuce.core.tracing.MicrometerTracing](https://github.com/redis/lettuce/blob/main/src/main/java/io/lettuce/core/tracing/MicrometerTracing.java) under Micrometer Observation. The issue is that this results in significant changes to the labels used in the metrics. Since these labels affect Grafana dashboards used across multiple products, updating them can be cumbersome.

For this reason, I would prefer to migrate to Micrometer Observation for tracing, while continuing to use MicrometerCommandLatencyRecorder for metrics, at least temporarily.

Currently, Micrometer Observation does not support disabling specific ObservationHandlers for a particular context. I would like to request the addition of such a feature. 

## Suggestion

For instance, adding a method like this to the ObservationConvention interface might be a possible approach:

```java
List<Class<?>> skipHandlers();
```

Would it be possible to consider implementing this feature?

**Comment by jonatan-ivanov** [comment]
> However, for legacy code that has already implemented metrics and tracing using older libraries such as Micrometer and Brave, transitioning entirely to Micrometer Observation can be challenging.

We thought about this use-case and we have components that can help you in such transitioning periods (see later).

>Currently, Micrometer Observation does not support disabling specific ObservationHandlers for a particular context.

It does, the [`ObservationHandler`]((https://github.com/micrometer-metrics/micrometer/blob/9917c67363f4507291b407edea57aebd2508b950/micrometer-observation/src/main/java/io/micrometer/observation/ObservationHandler.java#L89C1-L95C58)) has support for this:
```java
    /**
     * Tells the registry whether this handler should be applied for a given
     * {@link Observation.Context}.
     * @param context an {@link Observation.Context}
     * @return {@code true} when this handler should be used
     */
    boolean supportsContext(Observation.Context context);
```

>The issue is that this results in significant changes to the labels used in the metrics. Since these labels affect Grafana dashboards used across multiple products, updating them can be cumbersome.

We understand that, that's why we provide you these components:
- `ObservationPredicate`: You can disable entire `Observation`s based on name and the `Context`
- `ObservationFilter`: Simple way to modify `Observation`s (but this has a few limitations).
- `ObservationConvention`: You can redefine the `Observation`s what the instrumentation creates for you, this is somewhat similar to `ObservationFilter` in a sense that you can modify the `Observation`s that the instrumentation creates but without the limitations.

In your case what I would do is utilizing the `ObservationFilter` or the `ObservationConvention` and modify the `Observation`s so that the metrics outputs will be the same what your dashboards expect.

>For instance, adding a method like this to the ObservationConvention interface might be a possible approach:
```java
List<Class<?>> skipHandlers();
```
>Would it be possible to consider implementing this feature?

I think this is not a good idea from multiple perspectives:
- As I mentioned above, there is a mechanism that already exits for this: `ObservationHandler#supportsContext`.
- The instrumentation and the handler mechanism are separated for good reasons. No instrumentation (`ObservationConvention`) should know about any handlers, they should have zero control about what outputs will be created from the `Context`, they should only provide the data and should have no idea what will happen with that data later on.

So to recap, your two simplest options are:
1. Recommended: Modify the `Observation` using a custom `ObservationConvention` so that your dashboards will be happy.
2. Override the handlers you want and disable their outputs for certain `Context`s (`supportsContext`).

**Comment by be-hase** [comment]
I see your point. Given that ObservationHandler is often set globally in frameworks like Spring Boot, I agree that it would be more convenient to disable handlers locally using ObservationConvention in specific areas of interest. This would provide more flexibility and control.

That said, I also understand your reasoning, and since it can be handled through overriding, I'm okay with that approach.

**Comment by jonatan-ivanov** [comment]
Spring Boot does not set handlers globally, it sets them on a registry instance, there is no static global there, it is a single instance that is used throughout your app. You can replace this registry instance and register only the handlers you want (e.g.: a handler that only processes the observations you would like to be processed).
